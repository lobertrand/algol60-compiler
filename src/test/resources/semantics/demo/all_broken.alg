begin

    begin
        integer a, b;
        real a;
        ininteger(0, a);
        ininteger(0, b);

        begin
            comment Switch declaration;
            switch state := first, second, extra;
            boolean chooseFirst;

            first:;
                outstring(1, "first operation");
                goto switchEnd;

            second:;
                outstring(1, "second operation");
                goto switchEnd;

            chooseFirst := b > 5 \/ a + 3 /\ a+b < 40;
            comment If-expression;
            goto state[if chooseFirst then 1 else 2]
        end;

        begin
            switchEnd:
        end
    end;

    begin
        comment Array as parameter of procedure;
        integer procedure max(arr, a, b, extra);
        value arr, a, b; string array arr; integer a, b;
        begin
            integer i, result;
            result := arr[a];
            comment For-statement;
            for i := a+1 step 1 until b do
            begin
                comment If-statement;
                if arr[i] > result then
                    result := arr[i]
            end;
            max := result
        end;

        comment Array declaration and assignment;
        integer array a[1:3];
        a[0] := 6;
        a[2] := 8;
        a[4] := 3;

        comment Procedure call;
        outinteger(1, max(a, 1, 3))
    end;

    begin
        integer procedure getMessageNumber(grade);
        value grade; integer grade; integer extra;
        begin
            comment If-statement exhaustiveness check;
            if grade < 8 then
                getMessageNumber := 1
            else if grade < 14 then
                getMessageNumber := 2
            comment else
                getMessageNumber := 3 ;
        end;

        string a;
        a := 1;
        notDeclared();
        begin
            string array messages[a:3];
            comment string msg;

            messages[1] := "bad";
            messages[2] := "okay";
            messages[4] := "great";

            msg := messages[getMessageNumber(12.5)];
        end
    end;

    for a := 1, 2, 3 do
        myLabel:;

end
