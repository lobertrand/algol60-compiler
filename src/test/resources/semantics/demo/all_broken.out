Algol60> Lexical analysis successful
Algol60> Syntactic analysis successful
┌───────────────────────────────────────────────┐
│ level: 0, tableNumber: 0                      │
├───────────────────────────────────────────────┤
│ Procedure: void outstring(integer, string)    │
│ Procedure: void outboolean(integer, boolean)  │
│ Procedure: void outinteger(integer, integer)  │
│ Procedure: void outreal(integer, real)        │
│ Procedure: void inreal(integer, real)         │
│ Procedure: void ininteger(integer, integer)   │
│ Procedure: integer entier(real)               │
└───────────────────────────────────────────────┘
    ┌──────────────────────────┐
    │ level: 1, tableNumber: 1 │
    ├──────────────────────────┤
    └──────────────────────────┘
        ┌──────────────────────────┐
        │ level: 2, tableNumber: 2 │
        ├──────────────────────────┤
        │ Variable: integer a      │
        │ Variable: integer b      │
        └──────────────────────────┘
            ┌──────────────────────────────────────┐
            │ level: 3, tableNumber: 3             │
            ├──────────────────────────────────────┤
            │ Switch: state [first, second, extra] │
            │ Variable: boolean chooseFirst        │
            │ Label: first                         │
            │ Label: second                        │
            └──────────────────────────────────────┘
            ┌──────────────────────────┐
            │ level: 3, tableNumber: 4 │
            ├──────────────────────────┤
            │ Label: switchEnd         │
            └──────────────────────────┘
        ┌────────────────────────────────────────────────────────────────────┐
        │ level: 2, tableNumber: 5                                           │
        ├────────────────────────────────────────────────────────────────────┤
        │ Procedure: integer max(string array, integer, integer, undefined)  │
        │ Array: integer array a[1:3]                                        │
        └────────────────────────────────────────────────────────────────────┘
            ┌─────────────────────────────────────────┐
            │ level: 3, tableNumber: 6                │
            ├─────────────────────────────────────────┤
            │ Array: string array arr[] (value param) │
            │ Variable: integer a (value param)       │
            │ Variable: integer b (value param)       │
            │ Variable: undefined extra (ref. param)  │
            │ Variable: integer max (proc. result)    │
            │ Variable: integer i                     │
            │ Variable: integer result                │
            └─────────────────────────────────────────┘
                ┌──────────────────────────┐
                │ level: 4, tableNumber: 7 │
                ├──────────────────────────┤
                └──────────────────────────┘
        ┌───────────────────────────────────────────────┐
        │ level: 2, tableNumber: 8                      │
        ├───────────────────────────────────────────────┤
        │ Procedure: integer getMessageNumber(integer)  │
        │ Variable: string a                            │
        └───────────────────────────────────────────────┘
            ┌───────────────────────────────────────────────────┐
            │ level: 3, tableNumber: 9                          │
            ├───────────────────────────────────────────────────┤
            │ Variable: integer grade (value param)             │
            │ Variable: integer getMessageNumber (proc. result) │
            └───────────────────────────────────────────────────┘
            ┌────────────────────────────────────┐
            │ level: 3, tableNumber: 10          │
            ├────────────────────────────────────┤
            │ Array: string array messages[?:3]  │
            └────────────────────────────────────┘

Algol60> Symbol redeclaration at line 5: Variable 'a' is already declared in scope
     5 |         real a;
       |              ^
Algol60> Symbol not declared at line 11: Label 'extra' is not declared
    11 |             switch state := first, second, extra;
       |                                            ^
Algol60> Symbol not declared at line 16: Label 'switchEnd' is not declared
    16 |                 goto switchEnd;
       |                      ^
Algol60> Symbol not declared at line 20: Label 'switchEnd' is not declared
    20 |                 goto switchEnd;
       |                      ^
Algol60> Type mismatch at line 22: Cannot perform boolean operation on operands of types integer and boolean
    22 |             chooseFirst := b > 5 \/ a + 3 /\ a+b < 40;
       |                                           ^
Algol60> Parameter mismatch at line 34: Parameter 'extra' has no type specified in specification part of procedure 'max'
    34 |         integer procedure max(arr, a, b, extra);
       |                                          ^
Algol60> Type mismatch at line 38: Cannot assign a string to 'result' of type integer
    38 |             result := arr[a];
       |             ^
Algol60> Type mismatch at line 43: Cannot perform arithmetic operation on operands of types string and integer
    43 |                 if arr[i] > result then
       |                           ^
Algol60> Type mismatch at line 44: Cannot assign a string to 'result' of type integer
    44 |                     result := arr[i]
       |                     ^
Algol60> Index out of bounds at line 51: Array index '0' is out of bounds in 'a' with bounds 1:3
    51 |         a[0] := 6;
       |           ^
Algol60> Index out of bounds at line 53: Array index '4' is out of bounds in 'a' with bounds 1:3
    53 |         a[4] := 3;
       |           ^
Algol60> Parameter mismatch at line 56: Procedure 'max' expects 4 parameters but received 3
    56 |         outinteger(1, max(a, 1, 3))
       |                       ^
Algol60> Missing return at line 60: Procedure 'getMessageNumber' may never return an integer
    60 |         integer procedure getMessageNumber(grade);
       |                           ^
Algol60> Parameter mismatch at line 61: 'extra' is not a parameter of procedure 'getMessageNumber'
    61 |         value grade; integer grade; integer extra;
       |                                             ^
Algol60> Type mismatch at line 73: Cannot assign an integer to 'a' of type string
    73 |         a := 1;
       |         ^
Algol60> Symbol not declared at line 74: Procedure 'notDeclared' is not declared
    74 |         notDeclared();
       |         ^
Algol60> Type mismatch at line 76: Bounds must be of type integer or real but bound start is a string
    76 |             string array messages[a:3];
       |                                   ^
Algol60> Index out of bounds at line 81: Array index '4' is out of bounds in 'messages' with bounds ?:3
    81 |             messages[4] := "great";
       |                      ^
Algol60> Symbol not declared at line 83: Variable 'msg' is not declared
    83 |             msg := messages[getMessageNumber(12.5)];
       |             ^
Algol60> Type mismatch at line 83: Parameter #1 of procedure 'getMessageNumber' expects an integer but it received a real
    83 |             msg := messages[getMessageNumber(12.5)];
       |                                              ^
Algol60> Symbol not declared at line 87: Variable 'a' is not declared
    87 |     for a := 1, 2, 3 do
       |         ^
Algol60> Unreachable statement at line 88: Cannot declare a label inside a one-line for loop
    88 |         myLabel:;
       |         ^
Algol60> Terminated with 22 semantic exceptions
